Add a resource reference generator

ptgott opened the pull request:
Closes #16948
Closes #9949

Implements RFD 130.

See the included README for information about how the generator works.

This implementation breaks from, and updates, RFD 130:

- The reference includes separate pages for dynamic resources in order to make it easier to read. Edit the RFD to be less specific about the output template so this isn't wedded to implementation details.
- The generator uses a YAML configuration file. This is because the config ended up including more fields than specified in the RFD, so separating the config from the source made sense to give users less noise to deal with.
- The generator requires that a user specify the resources to include in the reference docs. This makes it easier to start with a smaller, more manageable set of resource reference docs.
- Support RFD 153 in the reference generator: RFD 153 resource types include unexported fields and include a `Metadata` field from a different package than legacy resources types. They also include the `Metadata` field with a pointer. Adjust the generator to ignore unexported fields in struct declarations, and to treat pointer field types the same way as value field types.
- Remove the description of example YAML delimiters. These are no longer viable, as we maintain several documentation generators that extract comments from the source. Instead, if we can't generate example YAML, show a comment in the YAML example advising the reader to read the description of a field.

Other changes:

 - Add a Make target for generating dynamic resource reference documentation.
 - Generate reference docs for Bot, RoleV6, OIDCConnectorV3, and SAMLConnectorV2.

ptgott committed:
Add a resource reference generator

Closes #16948
Closes #9949

Implements RFD 130.

See the included README for information about how the generator works.

This implementation breaks from, and updates, RFD 130:

- The reference includes separate pages for dynamic resources in order
  to make it easier to read. Edit the RFD to be less specific about the
  output template so this isn't wedded to implementation details.
- The generator uses a YAML configuration file. This is because the
  config ended up including more fields than specified in the RFD, so
  separating the config from the source made sense to give users less
  noise to deal with.
- The generator requires that a user specify the resources to include in
  the reference docs. This makes it easier to start with a smaller, more
  manageable set of resource reference docs.
- Support RFD 153 in the reference generator: RFD 153 resource types
  include unexported fields and include a `Metadata` field from a
  different package than legacy resources types. They also include the
  `Metadata` field with a pointer. Adjust the generator to ignore
  unexported fields in struct declarations, and to treat pointer field
  types the same way as value field types.
- Remove the description of example YAML delimiters. These are no longer
  viable, as we maintain several documentation generators that extract
  comments from the source. Instead, if we can't generate example YAML,
  show a comment in the YAML example advising the reader to read the
  description of a field.

Other changes:

 - Add a Make target for generating dynamic resource reference
   documentation, plus a target for checking that the resource reference
   docs are up to date. Run the checker in the `Lint (Proto)` job.
 - Generate reference docs for Bot, RoleV6, OIDCConnectorV3, and
   SAMLConnectorV2.

zmb3 commented on build.assets/tooling/cmd/resource-ref-generator/reference/reference.go:
```diff
@@ -0,0 +1,154 @@
+		// decl is a dynamic resource type, so get data for the type and
+		// its dependencies.
+		entries, err := resource.ReferenceDataFromDeclaration(decl, sourceData.TypeDecls)
+		if errors.Is(err, resource.NotAGenDeclError{}) {
```
I would test this and make sure it does what you think. You might need to use `errors.As` here instead.

zmb3 reviewed (approved):
Haven't tried running it myself, but assuming it works I'm happy to get this in and start using it. We can refine it as we go.

ptgott committed:
Clean up the resource reference generator

- Remove an unneeded import.
- Use slices.SortFunc instead of sort.Sort for sorting reference entry
  fields.
- Change `errors.Is(err, NotAGenDeclError{})` calls to use `errors.As`.
  While `NotAGenDeclError{}` structs are always empty, using `errors.As`
  ensures the calls still do the same thing if we add fields to the
  type.
